File: java/src/com/aes/AesCrypt.java
package com.aes;

import java.nio.charset.Charset;
import java.util.Arrays;
import java.util.Random;
import java.security.MessageDigest;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * 针对org.apache.commons.codec.binary.Base64，
 * 需要导入架包commons-codec-1.9（或commons-codec-1.8等其他版本）
 * 官方下载地址：http://commons.apache.org/proper/commons-codec/download_codec.cgi
 */
import org.apache.commons.codec.binary.Base64;

/**
 * 提供基于AES-CBC算法的加解密接口
 *
 * 说明：异常java.security.InvalidKeyException:illegal Key Size的解决方案
 * <ol>
 * 	<li>在官方网站下载JCE无限制权限策略文件（JDK7的下载地址：
 *      http://www.oracle.com/technetwork/java/javase/downloads/jce-7-download-432124.html</li>
 * 	<li>下载后解压，可以看到local_policy.jar和US_export_policy.jar以及readme.txt</li>
 * 	<li>如果安装了JRE，将两个jar文件放到%JRE_HOME%\lib\security目录下覆盖原来的文件</li>
 * 	<li>如果安装了JDK，将两个jar文件放到%JDK_HOME%\jre\lib\security目录下覆盖原来文件</li>
 * </ol>
 */
public class AesCrypt {
	static Charset CHARSET = Charset.forName("utf-8");
	Base64 base64 = new Base64();
	byte[] aesKey;
	String token;

	/**
	 * 构造函数
	 * @param token 签名校验Token
	 * @param encodingAesKey AES加解密Key
	 *
	 * @throws AesException 执行失败，请查看该异常的错误码和具体的错误信息
	 */
	public AesCrypt(String token, String encodingAesKey) throws AesException {
		if (encodingAesKey.length() != 43) {
			throw new AesException(AesException.IllegalAesKey);
		}

		this.token = token;
		aesKey = Base64.decodeBase64(encodingAesKey + "=");
	}

	// 生成4个字节的网络字节序
	byte[] getNetworkBytesOrder(int sourceNumber) {
		byte[] orderBytes = new byte[4];
		orderBytes[3] = (byte) (sourceNumber & 0xFF);
		orderBytes[2] = (byte) (sourceNumber >> 8 & 0xFF);
		orderBytes[1] = (byte) (sourceNumber >> 16 & 0xFF);
		orderBytes[0] = (byte) (sourceNumber >> 24 & 0xFF);
		return orderBytes;
	}

	// 还原4个字节的网络字节序
	int recoverNetworkBytesOrder(byte[] orderBytes) {
		int sourceNumber = 0;
		for (int i = 0; i < 4; i++) {
			sourceNumber <<= 8;
			sourceNumber |= orderBytes[i] & 0xff;
		}
		return sourceNumber;
	}

	// 随机生成16位字符串
	String getRandomStr() {
		String base = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
		Random random = new Random();
		StringBuffer sb = new StringBuffer();
		for (int i = 0; i < 16; i++) {
			int number = random.nextInt(base.length());
			sb.append(base.charAt(number));
		}
		return sb.toString();
	}

	/**
	 * 用SHA1算法生成安全签名
	 * @param token 票据
	 * @param text 密文
	 * @return 安全签名
	 * @throws AesException
	 */
	String getSHA1(String token, String text) throws AesException {
		try {
			String[] array = new String[] { token, text };
			StringBuffer sb = new StringBuffer();
			// 字符串排序
			Arrays.sort(array);
			for (int i = 0; i < 2; i++) {
				sb.append(array[i]);
			}
			String str = sb.toString();
			// SHA1签名生成
			MessageDigest md = MessageDigest.getInstance("SHA-1");
			md.update(str.getBytes());
			byte[] digest = md.digest();

			StringBuffer hexstr = new StringBuffer();
			String shaHex = "";
			for (int i = 0; i < digest.length; i++) {
				shaHex = Integer.toHexString(digest[i] & 0xFF);
				if (shaHex.length() < 2) {
					hexstr.append(0);
				}
				hexstr.append(shaHex);
			}
			return hexstr.toString();
		} catch (Exception e) {
			e.printStackTrace();
			throw new AesException(AesException.ComputeSignatureError);
		}
	}

	/**
	 * 对明文进行加密
	 * 
	 * @param text 需要加密的明文
	 * @return 加密后base64编码的字符串
	 * @throws AesException aes加密失败
	 */
	public String encrypt(String text) throws AesException {
		String randomStr = getRandomStr();
		ByteGroup byteCollector = new ByteGroup();
		byte[] randomStrBytes = randomStr.getBytes(CHARSET);
		byte[] textBytes = text.getBytes(CHARSET);
		byte[] networkBytesOrder = getNetworkBytesOrder(textBytes.length);

		// randomStr + networkBytesOrder + text
		byteCollector.addBytes(randomStrBytes);
		byteCollector.addBytes(networkBytesOrder);
		byteCollector.addBytes(textBytes);

		// ... + pad: 使用自定义的填充方式对明文进行补位填充
		byte[] padBytes = PKCS7Encoder.encode(byteCollector.size());
		byteCollector.addBytes(padBytes);

		// 获得最终的字节流, 未加密
		byte[] unencrypted = byteCollector.toBytes();

		try {
			// 设置加密模式为AES的CBC模式
			Cipher cipher = Cipher.getInstance("AES/CBC/NoPadding");
			SecretKeySpec keySpec = new SecretKeySpec(aesKey, "AES");
			IvParameterSpec iv = new IvParameterSpec(aesKey, 0, 16);
			cipher.init(Cipher.ENCRYPT_MODE, keySpec, iv);

			// 加密
			byte[] encrypted = cipher.doFinal(unencrypted);

			// 使用BASE64对加密后的字符串进行编码
			String base64Encrypted = base64.encodeToString(encrypted);

			return base64Encrypted;
		} catch (Exception e) {
			e.printStackTrace();
			throw new AesException(AesException.EncryptAESError);
		}
	}

	/**
	 * 对密文进行解密.
	 * 
	 * @param encrypted 需要解密的密文
	 * @return 解密得到的明文
	 * @throws AesException aes解密失败
	 */
	public String decrypt(String encrypted) throws AesException {
		byte[] deciphered;
		try {
			// 设置解密模式为AES的CBC模式
			Cipher cipher = Cipher.getInstance("AES/CBC/NoPadding");
			SecretKeySpec key_spec = new SecretKeySpec(aesKey, "AES");
			IvParameterSpec iv = new IvParameterSpec(Arrays.copyOfRange(aesKey, 0, 16));
			cipher.init(Cipher.DECRYPT_MODE, key_spec, iv);

			// 使用BASE64对密文进行解码
			byte[] bytes = Base64.decodeBase64(encrypted);

			// 解密
			deciphered = cipher.doFinal(bytes);
		} catch (Exception e) {
			e.printStackTrace();
			throw new AesException(AesException.DecryptAESError);
		}

		String content;
		try {
			// 去除补位字符
			byte[] bytes = PKCS7Encoder.decode(deciphered);

			// 分离16位随机字符串，网络字节序
			byte[] networkOrder = Arrays.copyOfRange(bytes, 16, 20);

			int length = recoverNetworkBytesOrder(networkOrder);

			content = new String(Arrays.copyOfRange(bytes, 20, 20 + length), CHARSET);
		} catch (Exception e) {
			e.printStackTrace();
			throw new AesException(AesException.IllegalBuffer);
		}

		return content;

	}

	/**
	 * 生成安全签名
	 * 
	 * @param encrypted 加密后的密文
	 *
	 * @return 签名字符串
	 * @throws AesException 执行失败，请查看该异常的错误码和具体的错误信息
	 */
	public String getSignature(String encrypted) throws AesException {

		String signature = getSHA1(token, encrypted);

		// System.out.println("发送给平台的签名是: " + signature[1].toString());
		return signature;
	}

}

File: nodejs/AesCrypt.js
/**
 * Nodejs版的加解密示例代码
 *
 */

'use strict';

const crypto = require('crypto');

/**
 * 提供基于PKCS7算法的加解密接口
 *
 */
let PKCS7Encoder = {};

/**
 * 删除解密后明文的补位字符
 *
 * @param {String} text 解密后的明文
 */
PKCS7Encoder.decode = function (text) {
    let pad = text[text.length - 1];
    if (pad < 1 || pad > 32) {
        pad = 0;
    }
    return text.slice(0, text.length - pad);
};

/**
 * 对需要加密的明文进行填充补位
 *
 * @param {String} text 需要进行填充补位操作的明文
 */
PKCS7Encoder.encode = function (text) {
    let blockSize = 32;
    let textLength = text.length;
    let amountToPad = blockSize - (textLength % blockSize);//计算需要填充的位数
    let result = new Buffer(amountToPad);
    result.fill(amountToPad);
    return Buffer.concat([text, result]);
};

/**
 * 提供基于AES-CBC算法的加解密接口
 *
 * @param {String} token 签名校验Token
 * @param {String} encodingAESKey AES加解密Key
 */
let AesCrypt = function (token, encodingAESKey) {
    if (!token || !encodingAESKey) {
        throw new Error('please check arguments');
    }
    if (encodingAESKey.length !== 43) {
        throw new Error('encodingAESKey invalid');
    }
    this.token = token;
    let AESKey = new Buffer(encodingAESKey + '=', 'base64');
    // if (AESKey.length !== 32) {
    //     throw new Error('encodingAESKey invalid');
    // }
    this.key = AESKey;
    this.iv = AESKey.slice(0, 16);
};

/**
 * 生成安全签名
 *
 * @param {String} encrypted 加密后的密文
 * @return {String}
 */
AesCrypt.prototype.getSignature = function(encrypted) {
    let shasum = crypto.createHash('sha1');
    let arr = [this.token, encrypted].sort();
    shasum.update(arr.join(''));
    return shasum.digest('hex');
};

/**
 * 对明文进行加密
 *
 * @param {String} text 待加密的明文
 * @return {String}
 */
AesCrypt.prototype.encrypt = function (text) {
    // 算法：AES_Encrypt[random(16B) + msgLen(4B) + msg]
    // 获取16B的随机字符串
    let randomStr = crypto.pseudoRandomBytes(16);
    let msg = new Buffer(text);
    // 获取4B的内容长度的网络字节序
    let msgLen = new Buffer(4);
    msgLen.writeUInt32BE(msg.length, 0);
    let collector = Buffer.concat([randomStr, msgLen, msg]);
    // 对明文进行补位操作
    let encoded = PKCS7Encoder.encode(collector);
    // 创建加密对象，AES采用CBC模式，数据采用PKCS#7填充；IV初始向量大小为16字节，取AESKey前16字节
    let cipher = crypto.createCipheriv('aes-256-cbc', this.key, this.iv);
    cipher.setAutoPadding(false);
    let cipheredMsg = Buffer.concat([cipher.update(encoded), cipher.final()]);
    // 返回加密数据的base64编码
    return cipheredMsg.toString('base64');
};

/**
 * 对密文进行解密
 *
 * @param {String} encrypted 待解密的密文
 * @return {String}
 */
AesCrypt.prototype.decrypt = function(encrypted) {
    let decipher = crypto.createDecipheriv('aes-256-cbc', this.key, this.iv);// 创建解密对象，AES采用CBC模式，数据采用PKCS#7填充；IV初始向量大小为16字节，取AESKey前16字节
    decipher.setAutoPadding(false);
    let deciphered = Buffer.concat([decipher.update(encrypted, 'base64'), decipher.final()]);
    deciphered = PKCS7Encoder.decode(deciphered);
    // 算法：AES_Encrypt[random(16B) + msg_len(4B) + msg]
    // 去除16位随机数
    let content = deciphered.slice(16);
    let length = content.slice(0, 4).readUInt32BE(0);
    let text = content.slice(4, length + 4).toString();
    return text;
};

module.exports = AesCrypt;
